src/useGet.tsx

```diff
@@ -0,0 +1,182 @@
+    const { data, responseError } = await processResponse(response);
+
+    if (signal.aborted) {
+      return;
+    } else if (!response.ok || responseError) {
+      const error = {
+        message: `Failed to fetch: ${response.status} ${response.statusText}${responseError ? " - " + data : ""}`,
+        data,
+        status: response.status,
+      };
+
+      setState({ ...state, loading: false, error });
+
+      if (!props.localErrorOnly && context.onError) {
+        context.onError(error, () => _fetchData(props, state, setState, context, abortControllers), response);
+      }
+    } else {
+      setState({ ...state, loading: false, data: resolve(data) });
+    }
+  } catch (e) {
+    setState({
+      ...state,
+      loading: false,
+      error: {
+        message: `Failed to fetch: ${e.message}`,
+        data: e.message,
+      },
+    });
+  }
+}
+
+export function useGet<TData = any, TError = any, TQueryParams = { [key: string]: any }>(
+  props: UseGetProps<TData, TQueryParams>,
+) {
+  const context = useContext(Context);
+
+  const fetchData = useCallback<typeof _fetchData>(
+    typeof props.debounce === "object"
+      ? debounce<typeof _fetchData>(_fetchData, props.debounce.wait, props.debounce.options)
+      : typeof props.debounce === "number"
+      ? debounce<typeof _fetchData>(_fetchData, props.debounce)
+      : props.debounce
+      ? debounce<typeof _fetchData>(_fetchData)
+      : _fetchData,
+    [props.debounce],
+  );
+
+  const [state, setState] = useState<GetState<TData, TError>>({
+    data: null,
+    response: null,
+    loading: !props.lazy,
+    error: null,
+  });
+
+  const abortControllers = useRef([new AbortController()]);
+
+  const prevProps = useRef<UseGetProps<TData, TQueryParams>>();
+  useEffect(() => {
+    const { base, path, resolve, queryParams } = prevProps.current || {
+      base: null,
+      path: null,
+      resolve: null,
+      queryParams: null,
+    };
+
+    if (
+      base !== props.base ||
+      path !== props.path ||
+      // both `resolve` props need to _exist_ first, and then be equivalent.
+      (resolve && props.resolve && resolve.toString() !== props.resolve.toString()) ||
+      !isEqual(queryParams, props.queryParams)
+    ) {
+      if (!props.lazy) {
+        fetchData(props, state, setState, context, abortControllers.current);
+      }
+    }
+
+    // Save props for later comparison
+    prevProps.current = props;
+
+    return () => abortControllers.current.forEach(i => i.abort());
+  }, [props.path, props.base, props.resolve, props.queryParams]);
+
+  return {
+    ...state,
+    absolutePath: url.resolve(
+      props.base || context.base,
+      props.queryParams ? `${props.path}?${qs.stringify(props.queryParams)}` : props.path,
+    ),
+    refetch: (options: Partial<Omit<UseGetProps<TData, TQueryParams>, "lazy">> = {}) =>
+      fetchData({ ...props, ...options }, state, setState, context, abortControllers.current),
```

Comment:
`state` can get stale here. Suggestion is to use `useReducer` see https://twitter.com/kentcdodds/status/1102647400915582981 or https://codesandbox.io/s/71r1rqnvr6

Reply:
Since this function is directly render, I don't see how 😅

Reply:
Let's deal with the issue of stale state if:

- We can clearly produce a failing test/reproduce it.
- It is an _actual_ issue in one of our projects instead of a future, conceptual issue that _may not exist_.

If the claim of stale state can be substantiated, let's substantiate it and fix it. If not, let's merge this now.